lib/src/model_utils.dart

```diff
@@ -166,3 +169,150 @@ String crossdartifySource(
   }
   return newSource;
 }
+
+/// An UnmodifiableListView that computes equality and hashCode based on the
+/// equality and hashCode of its contained objects.
+class HashableList extends UnmodifiableListView<dynamic> {
+  HashableList(Iterable<dynamic> iterable) : super(iterable);
+
+  @override
+  bool operator ==(other) {
+    if (this.length == other.length) {
+      for (var index = 0; index < length; ++index) {
+        if (this[index] != other[index]) return false;
+      }
+      return true;
+    }
+    return false;
+  }
+
+  @override
+  get hashCode => hashObjects(this);
+}
+
+/// A type alias for [Tuple2]`<Function, HashableList>`.
+class _MemoKey extends Tuple2<Function, HashableList> {
+  _MemoKey(Function f, HashableList l) : super(f, l) {}
+}
+
+/// Extend or use as a mixin to track object-specific cached values, or
+/// instantiate directly to track other values.
+class Memoizer {
+  /// Map of a function and its positional parameters (if any), to a value.
+  Map<_MemoKey, dynamic> _memoizationTable;
+
+  Memoizer() {
+    invalidateMemos();
+  }
+
+  /// Reset the memoization table, forcing calls of the underlying functions.
+  void invalidateMemos() {
+    _memoizationTable = new Map();
+  }
+
+  // Never write this if statement again.
+  R _getOrUpdateCache<R>(R Function() f, _MemoKey key) {
+    if (!_memoizationTable.containsKey(key)) {
+      _memoizationTable[key] = f();
+    }
+    return _memoizationTable[key];
+  }
+
+  /// Calls and caches the return value of [f]() if not in the cache, then
+  /// returns the cached value of [f]().
```

Comment:
Might want to add a comment explaining that `f` should be a tear-off of a method, not an inline closure, in order to guarantee that on subsequent calls the value of `f` will compare equal.  For example, clients should do this:
```dart
_doExpensiveThing() { ... }
memoizedExpensiveThing() => memoized(_doExpensiveThing);
```
rather than this:
```dart
memoizedExpensiveThing() => memoized(() { ... });
```

And similar comments in `memoized1`, `memoized2`, etc.

Reply:
Gave an outline of this idea in the class docs.